raiden/tests/unit/test_channelstate.py

```diff
@@ -1077,3 +1078,50 @@ def test_channelstate_withdraw():
     )
     iteration = channel.handle_channel_closed(channel_state, state_change)
     assert must_contain_entry(iteration.events, ContractSendChannelWithdraw, {})
+
+
+def test_channel_withdraw_must_not_change_merkletree():
+    our_model1, _ = create_model(70)
+    partner_model1, privkey2 = create_model(100)
+    channel_state = create_channel_from_models(our_model1, partner_model1)
+    payment_network_identifier = factories.make_address()
+
+    lock_amount = 10
+    lock_expiration = 100
+    lock_secret = sha3(b'test_channelstate_mediatedtransfer_overspent')
+    lock_hashlock = sha3(lock_secret)
+    lock = HashTimeLockState(
+        lock_amount,
+        lock_expiration,
+        lock_hashlock,
+    )
+
+    nonce = 1
+    transferred_amount = 0
+    receive_mediatedtransfer = make_receive_transfer_mediated(
+        channel_state,
+        privkey2,
+        nonce,
+        transferred_amount,
+        lock,
+    )
+
+    is_valid, msg = channel.handle_receive_mediatedtransfer(
+        channel_state,
+        receive_mediatedtransfer,
+    )
+    assert is_valid, msg
+
+    assert merkleroot(channel_state.partner_state.merkletree) == lock.lockhash
+
+    withdraw = ContractReceiveChannelWithdraw(
+        payment_network_identifier,
+        channel_state.token_address,
+        channel_state.identifier,
+        lock_secret,
+        channel_state.our_state.address,
+    )
+    iteration = channel.handle_channel_withdraw(channel_state, withdraw)
```

Comment:
The channel is not closed but the state change is handled without an error. We have a few things we could do here:

- Keep the behavior as it is right now
- Assert on the channel state, to enforce the closed state on it
- Guard the code with a state check, and ignore the state change if the channel is not closed

I'm not sure which is best.

Reply:
Withdrawing when the channel is not closed should be an errorm right? So option (3) sounds best?

Reply:
We can definitely go with option 3. But this case is really weird, the lock withdraw should come from a blockchain event, if the smart contract allowed it and the node thinks the channel is open, something is wrong. Perhaps we should add an event for `ProtocolWasBroken` or similar?

Reply:
We could do that but then what if this happened due to a state change being processed twice for whatever reason?